compiler/rustc_hir_analysis/src/collect/predicates_of.rs

```diff
@@ -330,6 +359,62 @@ fn gather_explicit_predicates_of(tcx: TyCtxt<'_>, def_id: LocalDefId) -> ty::Gen
     }
 }
 
+fn bidirectional_lifetime_predicates<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    item_def_id: LocalDefId,
+    bidir_outlives: impl Iterator<Item = (Lifetime, (LocalDefId, Symbol, Span))>,
+) -> Vec<(ty::Clause<'tcx>, Span)> {
+    let mut predicates = Vec::new();
+    let generics = tcx.generics_of(item_def_id.to_def_id());
+
+    bidirectional_lifetime_predicates_inner(
+        tcx,
+        item_def_id,
+        bidir_outlives,
+        generics,
+        &mut predicates,
+    );
+
+    predicates
+}
+
+fn bidirectional_lifetime_predicates_inner<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    item_def_id: LocalDefId,
+    bidir_outlives: impl Iterator<Item = (Lifetime, (LocalDefId, Symbol, Span))>,
+    generics: &Generics,
+    predicates: &mut Vec<(ty::Clause<'tcx>, Span)>,
+) {
+    let icx = ItemCtxt::new(tcx, item_def_id);
+
+    for (arg, (dup_def, name, span)) in bidir_outlives {
+        let orig_region = icx.astconv().ast_region_to_region(&arg, None);
+        if !matches!(orig_region.kind(), ty::ReEarlyBound(..)) {
```

Comment:
Maybe change the comment here to say something like "there is no late-bound lifetime to actually match up here, since the lifetime doesn't show up in the opaque's parent's substs"?